### DISCUSSION TITLE
Episode 1: Welcome to hell!

### DISCUSSION THREAD

**Discussion by DamienAmbrose** [Default-Cube-Studios/Movement-Template#15] [Devlogs]
# [Rant Warning] Input Crisis
> Pre-rant context:
> 1. The issue for the new input system ( #8 )
> 2. The idiot writing this ( @DamienAmbrose )

Let me set the scene.

Circa week ago, I was binge watching some Brackeys videos (RIP, you shall be missed) when I stumbled upon a tutorial about migrating to _Unity's new input system_.

I wasn't new to the pain and suffering of the legacy one. `GetAxis()` still haunts me.

> I still stand by the decision to migrate to the new version, but it'd be an understatement to say that it was going to be hard. In hindsight, I should have watched a tutorial or two before I started migrating.

So, what went wrong?

**Everything.**

Cue the exhibits!
## **Exhibit A:** 

So, imagine this. You decide to use a package released by UNITY THEMSELVES. You'd expect at least some coherent documentation or at least something to let you migrate quickly from the old input system, right?

**WRONG.**

The entire migration process is lengthy, and the documentation for it is either outdated, buggy, or doesn't tell you what you need to know.

For example, I tried looking for some docs for `"Control groups"` and what each one does. Guess what I found?

Nothing. Null. Zero. Zilch.

> A very special thank you to Mr. John Leonard French from _Game Dev Beginner_ for saving my soul by publishing a _very_ in-depth guide on how you use the dang thing.
https://gamedevbeginner.com/input-in-unity-made-easy-complete-guide-to-the-new-system

## **Exhibit B:**
Keeping your code modular is extremely important.

Heck, I'm supposed to be the person yelling at others for not following style guides or for writing redundant and unnecessary code.

But I've seen things, man.

The new input system is SO modular you need to write code to do SUCH. SIMPLE. STUFF.

For example, the input system fires functions when an action is disabled or reenabled. So, everything is Gucci then, right? Just write in the functionality and you'll be done!

**EXCEPT.**

**It's not.**

You need to write code. To disable. And enable. Your actions. Yourself.

[sighs] It's almost like the legacy system was more intuitive. 

(Remembers `GetAxis()`)

_Almost._

## **Exhibit C:**
NOTHING. WORKS.

Everything is SO buggy.

Clipping ( #5 ) used to be a problem, but now it's somehow gotten worse!

I'm clipping through the ground even when I'm idle!

The profiler hates me too now!

I'm seeing random spikes from the physics engine, the garbage collector, and some of my scripts, for no apparent reason!

I swear, if we have to roll back to the old input system just because of my incompetence, I'm going to punch myself in the face.

> Future Damien here. It turns out that the spikes from the profiler used to be there on previous builds too! I have literally nothin' to say at this point.

## **Exhibit D:**
The team behind the development of the input system had no idea how to handle input, so they just slapped on a bunch of ways to handle it!

Like, literally, there are 4 different ways to handle input that are built into the inspector.

You can even import stuff from `JSON` files.

It's like they had no idea what mortals like us wanted!

>This **isn't** to dunk on the people who worked on the project. They do great stuff and the input system isn't the worst; this is merely my opinion.

##
[sighs]
That felt good.

So, what are YOUR experiences with upgrading your projects?
Am I just stupid or is the new input system flawed?
Is there something I missed?

**Comment by ziyouxiaozi** [comment]
这是来自QQ邮箱的假期自动回复邮件。 您好，我已经收到你的邮件，我将在查看结束后，尽快给您回复。
